homeassistant/components/roomba/irobot_base.py

```diff
@@ -0,0 +1,218 @@
+"""Base class for iRobot devices."""
+import logging
+
+from homeassistant.components.vacuum import (
+    SUPPORT_BATTERY,
+    SUPPORT_LOCATE,
+    SUPPORT_PAUSE,
+    SUPPORT_RETURN_HOME,
+    SUPPORT_SEND_COMMAND,
+    SUPPORT_STATUS,
+    SUPPORT_STOP,
+    SUPPORT_TURN_OFF,
+    SUPPORT_TURN_ON,
+    VacuumDevice,
+)
+
+from . import roomba_reported_state
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+ATTR_CLEANING_TIME = "cleaning_time"
+ATTR_CLEANED_AREA = "cleaned_area"
+ATTR_ERROR = "error"
+ATTR_POSITION = "position"
+ATTR_SOFTWARE_VERSION = "software_version"
+
+# Commonly supported features
+SUPPORT_IROBOT = (
+    SUPPORT_BATTERY
+    | SUPPORT_PAUSE
+    | SUPPORT_RETURN_HOME
+    | SUPPORT_SEND_COMMAND
+    | SUPPORT_STATUS
+    | SUPPORT_STOP
+    | SUPPORT_TURN_OFF
+    | SUPPORT_TURN_ON
+    | SUPPORT_LOCATE
+)
+
+
+class IRobotBase(VacuumDevice):
+    """Base class for iRobot robots."""
+
+    def __init__(self, roomba, blid):
+        """Initialize the Roomba handler."""
+        self.vacuum = roomba
+        self.vacuum_state = roomba_reported_state(roomba)
+        self._blid = blid
+        self._name = self.vacuum_state.get("name")
+        self._version = self.vacuum_state.get("softwareVer")
+        self._sku = self.vacuum_state.get("sku")
+        self._cap_position = self.vacuum_state.get("cap", {}).get("pose") == 1
+
+        # Register callback function
+        self.vacuum.register_on_message_callback(self.on_message)
```

Comment:
Please move side effects out of `__init__`. It can be another function that is called in `async_setup_entry` after the entity is created